variables.tf

```
@@ -12,14 +12,32 @@ variable "create_vpc_endpoints" {
 
 variable "sg_egress_rules" {
   description = "Egress rules for the VPC Endpoint SecurityGroup(s). Set to empty list to disable default rules."
-  type        = list
-  default     = null
+  type = list(object({
+    description      = string
+    prefix_list_ids  = list(string)
+    from_port        = number
+    to_port          = number
+    protocol         = string
+    cidr_blocks      = list(string)
+    ipv6_cidr_blocks = list(string)
+    security_groups  = list(string)
+  }))
+  default = null
```

Thoughts on simplifying the dynamic block logic (eliminate the condition) by setting the default values here?

```suggestion
  default = [{
    description      = null
    prefix_list_ids  = null
    from_port        = 0
    to_port          = 0
    protocol         = "-1'
    cidr_blocks      = ["0.0.0.0/0"]
    ipv6_cidr_blocks = null
    security_groups  = null
  }
```

---

If I could, I absolutely would. But Terraform does not have a way to specify [partial defaults](https://github.com/hashicorp/terraform/issues/16517). 😢

---

Right, but I mean the defaults would be _instead of_ the locals. The default value here would be the full spec, with all attributes, using `null` where warranted...

---

Ah, got ya.

For the Egress rules I can do this. But for Ingress rules I cannot. I cannot set the default `cidr_blocks` value this way as it comes from a data source. Variables and their values are handled at parse time (i.e. during DAG generation) while the `data "awc_vpc"` block is handled during runtime.

I _could_ set `cidr_blocks` to `null` in the default but then I'd end up with more complex logic in the ingress dynamic block to decide when to use the variable's `cidr_block` vs. when to use the data lookup.

Should I go ahead and remove the `local.sg_egress_rules_default` and simplify at least that portion of things?

---

Kinda proposing just using 0.0.0.0 for the default ingress? I mean, there is some user-beware involved, but easy enough for the user to do the lookup and override on their end I think?

---

🤔 Hmm. I have an internal conflict going on about it but it's not a wrong decision. I like that it reduces complexity and thus the potential for bugs. I also like that as a user I am almost forced to be explicit about my choice.

My internal conflict revolves around being secure by default. I feel like it is akin to allowing SSH from `0.0.0.0/0` by default. I don't like it but many systems choose it (not just TF modules).

I'm on the fence. If it were a private project I would lean towards completely explicit usage -- default `cidr_blocks = null` and force the user to configure it. But I (more accurately, my company) would be the only consumer and I would have less need for "works by default".

For an open source project it feels more natural to provide a "works by default" experience.

---

> For an open source project it feels more natural to provide a "works by default" experience.

I agree with that. Maybe you could use 0.0.0.0 for the default but make it more visible in the `README.md`

---

I went ahead and got rid of both `locals` in favor of defaults for both Ingress and Egress. The docs generation already includes the default values but I did add a section at the top of the readme as well.

---

This works for me. I'll exercise it tomorrow morning...

---

Nice, thank you!

---

Morning was ambitious. Running it now. If clean, will merge and release as v4.0.0 due to the change in input types...